frontend/src/common/middlewares/refresher.js

```diff
@@ -0,0 +1,101 @@
+import { RestBuilder } from '../../public/utils/rest';
+import schema from '../schema';
+import {
+    createParamsForProjects,
+    urlForProjects,
+} from '../rest';
+import {
+    tokenSelector,
+    activeUserSelector,
+} from '../../common/selectors/auth';
+import {
+    setUserProjectsAction,
+} from '../../common/action-creators/domainData';
+
+
+export const START_REFRESH = 'refresh/START';
+export const STOP_REFRESH = 'refresh/STOP';
+
+export const startRefreshAction = loadCallback => ({
+    type: START_REFRESH,
+    loadCallback,
+});
+
+export const stopRefreshAction = () => ({
+    type: STOP_REFRESH,
+});
+
+
+class Refresher {
+    constructor(store) {
+        this.projectsRequest = this.createProjectsRequest(store);
+    }
+
+    createProjectsRequest = (store) => {
+        const projectsRequest = new RestBuilder()
+            .url(urlForProjects)
+            .params(() => {
+                const { access } = tokenSelector(store.getState());
+                return createParamsForProjects({ access });
+            })
+            .decay(0.3)
+            .maxRetryTime(3000)
+            .maxRetryAttempts(1)
+            .success((response) => {
+                try {
+                    schema.validate(response, 'projectsGetResponse');
+                    const { userId } = activeUserSelector(store.getState());
+                    store.dispatch(setUserProjectsAction({
+                        userId,
+                        projects: response.results,
+                    }));
+
+                    if (this.loadCallback) {
+                        this.loadCallback();
+                    }
+                } catch (er) {
+                    console.error(er);
+                }
+            })
+            .failure((response) => {
+                console.info('FAILURE:', response);
+                // TODO: logout and send to login screen
+            })
+            .fatal((response) => {
+                console.info('FATAL:', response);
+                // TODO: user couldn't be verfied screen
+            })
+            .build();
+        return projectsRequest;
+    }
+
+    load = (loadCallback) => {
+        this.stop();
+        this.loadCallback = loadCallback;
```

Comment:
The projectsRequest is not started here.
Why invoke stop here?

Reply:
Because the module doesn't necessarily know how it's being used.

I think it's best for every module that has a lifecycle and is exposed to others need to:

- when stopping: check if it has started
- when starting: assume restart and just stop before starting

Same goes for resource managers:

- make sure object exists before deleting
- delete before creating